### ISSUE TITLE
Consider some convenience changes for MVC options

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#6161]
It would be really nice to be able to write:

```c#
services.AddMvc(options => options.ModelBinderProviders.RemoveByType(typeof(ByteArrayModelBinderProvider));
```

or 

```c#
services.AddMvc().RemoveModelBinderProvider(typeof(ByteArrayModelBinder));
```

Instead of:

```c#
services.AddMvc(options => options.ModelBinderProviders.Remove(options.ModelBinderProviders.OfType<ByteArrayModelBinderProvider>().Single()));
```

See https://github.com/aspnet/Mvc/issues/6160#issuecomment-295296083

---- 
The first option would be expose a custom collection type, the second would be to just add more stuff to the `IMvcBuilder`. Either one of these could work for this scenario, but I think the builder approach could be to basically anything we want to make into a one liner.

**Comment by Eilon** [comment]
@jbagga - if this is super easy to add (+ tests), which I think it would be, let's sneak it into 2.0.0. If not, 2.1.0 is fine.

**Comment by henkmollema** [comment]
Which path should we take here? Is there a preference for extension methods on `IMvcBuilder` vs a custom collection type for model binder providers?

**Comment by henkmollema** [comment]
Right now I'm thinking of something like this:
```cs
public static IMvcBuilder RemoveModelBinderProvider<T>(this IMvcBuilder builder) where T : IModelBinderProvider
{
    builder.AddMvcOptions(o => o.ModelBinderProviders.Remove(o.ModelBinderProviders.OfType<T>().Single()));
    return builder;
}
```

Though I'm not sure if `builder.AddMvcOptions()` is right thing to use. Is there another way to get a hold of the options?

**Comment by davidfowl** [comment]
This is cleaner:

services.AddMvc(options => options.ModelBinderProviders.RemoveByType(typeof(ByteArrayModelBinderProvider));

**Comment by henkmollema** [comment]
That is indeed clean. However that requires a custom collection type.

**Comment by rynowak** [comment]
I think we'd prefer the custom collection types over extensions on the builder. This is how the filters convenience methods already work, so we should be consistent.

> That is indeed clean. However that requires a custom collection type.

I think we're ok with that. If you inherit from `Collection<T>` there's literally nothing you have to write except for the new methods.

**Comment by henkmollema** [comment]
Great, I'll do that.

Would the API look like this?
```cs
public class ModelBinderProviderCollection : Collection<IModelBinderProvider>
{
    RemoveByType<TModelBinderProvider>();
    RemoveByType(Type modelBinderProviderType);
}
```

**Comment by rynowak** [comment]
I think we'll need to call it `RemoveType`, since that's what's on `FormatterCollection`. https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/Formatters/FormatterCollection.cs#L36

So it would look like:
```
public class ModelBinderProviderCollection : Collection<IModelBinderProvider>
{
    RemoveType<TModelBinderProvider>();
    RemoveType(Type modelBinderProviderType);
}
```

So it looks like we should do this for:
- Conventions
- ModelBinderProviders
- ModelMetadataDetailsProviders 
- ModelValidatorProviders
- ValueProviderFactories 

And add the `RemoveType(Type type)` to formatter collection, which its missing from

see: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs

**Comment by rynowak** [comment]
Oh and as for where these types go, they belong with whatever `T` is. So `ModelBinderProviderCollection` -> `Microsoft.AspNetCore.Mvc.ModelBinding`

**Comment by henkmollema** [comment]
Sounds good. Does it makes sense to create a base class for such collection types? All their logic for removing is the same.

**Comment by henkmollema** [comment]
For the scope of this PR and 2.0.0: do we want this on all the types you mentioned or just model binder providers for now?

**Comment by rynowak** [comment]
We've been down that road before and I don't want to go down it again 😑 

If you want to try it and it's not gross, then sure. 

I still want to have a nominal collection type for each of these because that's the place where any tailored functionality goes (like filters have).

> For the scope of this PR and 2.0.0: do we want this on all the types you mentioned or just model binder providers for now?

We'll take as much as you want to give us. This bug stays open until I'm satisfied 😆 We weren't planning to do any of this work in 2.0.0 so it's a bonus for us.

**Comment by henkmollema** [comment]
Yeah I know, base classes can be evil. 😝 If there is not very much code duplication might as well go with individual classes.

I'll just get started and see where it ends. 👍

**Comment by rynowak** [comment]
Yeah. Wayyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy back when we had a sort of 'omni-options-item-collection' thing. It was pretty gross and made our lives better to delete. But we lost of the convenience things like this when we did it.

**Comment by davidfowl** [comment]
OptionsDescriptor 😄

**Comment by henkmollema** [comment]
I think I remember that, lol. 

Although this is a rather specific set of types, it might not be worth it. And even if we did, what name would it have? `RemovableTypeColletion<T>` 😆 ?

**Comment by rynowak** [comment]
We've done all we're going to do here for 2.0.0 but I want to leave this open to track future improvements.

**Comment by mkArtakMSFT** [comment]
@rynowak, is anything left here for us to handle during 2.2 ?

**Comment by dougbu** [comment]
@rynowak @mkArtakMSFT what "future improvements" is this issue tracking exactly?

**Comment by rynowak** [comment]
The changes here require binary breaking changes to define and return concrete collection types for these properties. We should evaluate this for 3.0 - but I don't think there's much we should do now.

**Comment by dougbu** [comment]
@mkArtakMSFT pushing this to 3.0 and removing my assignment based on @rynowak's month-old comment:
> I don't think there's much we should do now.
